activerecord/lib/active_record/connection_adapters/postgresql/oid/interval.rb

```diff
@@ -0,0 +1,42 @@
+require 'active_support/duration'
+
+module ActiveRecord
+  module ConnectionAdapters
+    module PostgreSQL
+      module OID # :nodoc:
+        class Interval < Type::Value # :nodoc:
+          def type
+            :interval
+          end
+
+          def type_cast_from_database(value)
+            if value.kind_of? ::String
+              ::ActiveSupport::Duration.parse(value)
+            else
+              super
+            end
+          end
+
+          def type_cast_from_user(value)
+            type_cast_from_database(value)
+          rescue ::ActiveSupport::Duration::ISO8601ParsingError
+            value # Allow user to supply raw string values in another formats supported by PostgreSQL
+          end
+
+          def type_cast_for_database(value)
+            case value
+              when ::ActiveSupport::Duration
+                value.iso8601(self.precision)
+              when ::Numeric
```

Comment:
Unfortunately some math with `Time`s results not in `Duration`s being returned, but `Float`s with number in seconds:
```
Time.current - (Time.current + 1.hour) # => -3600.00000453 (Float)
```
I think that we should be able to handle that.